File: ReactNativeApp/ios/Modules/SSVC/Instance/Detection/Person/PersonController.swift
//
//  PersonController.swift
//  jogo
//
//  Created by Mohsin on 18/10/2021.
//

import Foundation

public class PersonController: Controller {
  
  public var highestLeftAnkle: DetectionLocation?
  public var highestRightAnkle: DetectionLocation?
  public var lowestLeftHeel: DetectionLocation?
  public var lowestRightHeel: DetectionLocation?
  public var lowestHighLeftHeel: DetectionLocation?
  public var lowestHighRightHeel: DetectionLocation?
  
  let JUMP_LOOKBACK = 30
  let GROUND_LOOKBACK: Int = 6
  let HEAD_LOOKBACK: Int = 6
  let MIN_JUMP: Double = 0.08
  let MIN_JUMP_FRAME_DISTANCE: Int = 5
  
  var personState: PersonState?
  
  private var groundLine: Double = 0
  
  init(state: PersonState) {
    super.init(state: state)
    self.personState = state
  }
  
  public func didJump(lastJumpFrameId: Int) -> Int {
    let stime: Double = Date().getMilliseconds()
    if personState!.getInfoBlobArrayList().count() < JUMP_LOOKBACK { return 0 }
    var lastFrameId: Int
    do {
      lastFrameId = try personState!.getInfoBlobArrayList().getLast()!.frameId!
    } catch {
      Logger.shared.logError(logType: .ERROR, error: error)
      Profiler.get().profileTime(frameId: lastJumpFrameId, tag: "Person-\(#function)-false", delta: Date().getMilliseconds() - stime)
      return 0 }
    
    let leftHeels = personState!.leftLeg!.heel.getNDetectionLocations(locationCount: min(JUMP_LOOKBACK, lastFrameId - lastJumpFrameId))
    let rightHeels = personState!.rightLeg!.heel.getNDetectionLocations(locationCount: min(JUMP_LOOKBACK, lastFrameId - lastJumpFrameId))
    
    if leftHeels == nil || rightHeels == nil {
      Profiler.get().profileTime(frameId: lastFrameId, tag: "Person-\(#function)-false", delta: Date().getMilliseconds() - stime)
      return 0 }
    if leftHeels!.isEmpty() || rightHeels!.isEmpty() {
      Profiler.get().profileTime(frameId: lastFrameId, tag: "Person-\(#function)-false", delta: Date().getMilliseconds() - stime)
      return 0 }
    
    lowestLeftHeel = leftHeels!.getMin(comparator: { (det1, det2) -> Bool in
      det1.getY() < det2.getY()
    })
    lowestRightHeel = rightHeels!.getMin(comparator: { (det1, det2) -> Bool in
      det1.getY() < det2.getY()
    })
    let leftHighIndex: Int = leftHeels!.get().firstIndex { (detLoc) -> Bool in
      detLoc.frameId == lowestLeftHeel!.frameId
    }!
    let rightHighIndex: Int = rightHeels!.get().firstIndex { (detLoc) -> Bool in
      detLoc.frameId == lowestRightHeel!.frameId
    }!
    
    let leftAnkles = leftHeels?.get()[0 ..< leftHighIndex]
    let rightAnkles = rightHeels?.get()[0 ..< rightHighIndex]
    if leftAnkles == nil || rightAnkles == nil {
      Profiler.get().profileTime(frameId: lastFrameId, tag: "Person-\(#function)-false", delta: Date().getMilliseconds() - stime)
      return 0 }
    if leftAnkles!.isEmpty || rightAnkles!.isEmpty {
      Profiler.get().profileTime(frameId: lastFrameId, tag: "Person-\(#function)-false", delta: Date().getMilliseconds() - stime)
      return 0 }
    highestLeftAnkle = leftAnkles!.max(by: { (det1, det2) -> Bool in
      det1.getY() < det2.getY()
    })
    highestRightAnkle = rightAnkles!.max(by: { (det1, det2) -> Bool in
      det1.getY() < det2.getY()
    })
    
    let leftHighHeels = leftHeels?.get()[leftHighIndex + 1 ..< (leftHeels?.count())!]
    let rightHighHeels = rightHeels?.get()[rightHighIndex + 1 ..< (rightHeels?.count())!]
    if leftHighHeels == nil || rightHighHeels == nil {
      return 0 }
    lowestHighLeftHeel = leftHighHeels!.max(by: { (det1, det2) -> Bool in
      det1.getY() < det2.getY()
    })
    lowestHighRightHeel = rightHighHeels!.max(by: { (det1, det2) -> Bool in
      det1.getY() < det2.getY()
    })
    
    if lowestLeftHeel == nil || lowestRightHeel == nil || lowestHighLeftHeel == nil || lowestHighRightHeel == nil {
      Profiler.get().profileTime(frameId: lastFrameId, tag: "Person-\(#function)-false", delta: Date().getMilliseconds() - stime)
      return 0
    }
    
    if jumpConditionOne() {
      if (max(lowestHighLeftHeel!.getFrameId(), lowestHighRightHeel!.getFrameId()) - lastJumpFrameId) <= MIN_JUMP_FRAME_DISTANCE {
        return lastJumpFrameId
      } else {
        return max(lowestLeftHeel!.getFrameId(), lowestRightHeel!.getFrameId())
      }
    }
    Profiler.get().profileTime(frameId: lastFrameId, tag: "Person-\(#function)-true", delta: Date().getMilliseconds() - stime)
    return lastJumpFrameId
  }
  
  public func getGroundLine() -> Double {
    let leftAnkleLocations: UtilArrayList<DetectionLocation>? = personState!.leftLeg!.ankle.getNDetectionLocations(locationCount: GROUND_LOOKBACK)
    let rightAnkleLocations: UtilArrayList<DetectionLocation>? = personState!.rightLeg!.ankle.getNDetectionLocations(locationCount: GROUND_LOOKBACK)
    if leftAnkleLocations == nil || rightAnkleLocations == nil { return groundLine }
    let highLeft: Double = leftAnkleLocations!.getMax(comparator: { (det1, det2) -> Bool in
      det1.getY() < det2.getY()
    })!.getY()
    
    let highRight: Double = rightAnkleLocations!.getMax { (det1, det2) -> Bool in
      det1.getY() < det2.getY()
    }!.getY()
    groundLine = max(highLeft, highRight)
    return groundLine
  }
  
  public func getHeadLine() -> Double {
    let stime: Double = Date().getMilliseconds()
    let leftEyeLocations: UtilArrayList<DetectionLocation>? = personState!.face!.leftEye.getNDetectionLocations(locationCount: GROUND_LOOKBACK)
    let rightEyeLocations: UtilArrayList<DetectionLocation>? = personState!.face!.rightEye.getNDetectionLocations(locationCount: GROUND_LOOKBACK)
    if leftEyeLocations == nil || rightEyeLocations == nil { return -1 }
    let highLeft: Double = leftEyeLocations!.getMin(comparator: { (det1, det2) -> Bool in
      det1.getY() < det2.getY()
    })!.getY()
    
    let highRight: Double = rightEyeLocations!.getMin { (det1, det2) -> Bool in
      det1.getY() < det2.getY()
    }!.getY()
    let headLine = max(highLeft, highRight)
    Profiler.get().profileTime(frameId: -1, tag: "Person-\(#function)-true", delta: Date().getMilliseconds() - stime)
    return headLine
  }
  
  public func inFrame() -> Bool {
    return personState!.getDetectionSubClasses().allSatisfy({$0.inFrame()})
  }
  
  func jumpConditionOne() -> Bool {
    return (lowestLeftHeel!.getY() + MIN_JUMP) < highestLeftAnkle!.getY() &&
              (lowestRightHeel!.getY() + MIN_JUMP) < highestRightAnkle!.getY() &&
              (lowestHighLeftHeel!.getY() - MIN_JUMP) > lowestLeftHeel!.getY() &&
              (lowestHighRightHeel!.getY() - MIN_JUMP) > lowestRightHeel!.getY()
  }
  
  func changeCalibrationStatus(isCalibrated: Bool){
    personState?.isCalibrated = isCalibrated
  }
  
  public override func getDebugText() -> String {
    return "PersonController"
  }
  
  internal func getBodyPart(label: Posenet.BODYPART) -> ObjectDetectionState? {
    var bodyElement: ObjectDetectionState?
    self.personState?.bodyParts.forEach({ bodyPart in
      bodyPart.bodyElements.forEach({ element in
        if element.getLabel() == label.rawValue {
          bodyElement = element
        }
      })
    })
    return bodyElement
  }
  
  public override func cleanup() {
    super.cleanup()
    self.personState?.removeObservers()
    self.personState = nil
    self.highestLeftAnkle = nil
    self.highestRightAnkle = nil
    self.lowestLeftHeel = nil
    self.lowestRightHeel = nil
    self.lowestHighLeftHeel = nil
    self.lowestHighRightHeel = nil
  }
}

File: ReactNativeApp/android/app/src/main/java/com/components/detection/person/PersonController.java
package com.components.detection.person;

import com.components.detection.ObjectDetectionState;
import com.components.detection.person.bodyPartDetection.ArmState;
import com.location.DetectionLocation;
import com.logger.SLOG;
import com.ssvc.Controller;
import com.utils.SMath;
import com.utils.UtilArrayList;

public class PersonController extends Controller<PersonState> {

    protected static final float EXTENSION_THRESHOLD = 0.90f;
    final int GROUND_LOOKBACK = 6, HEAD_LOOKBACK = 6;
    public DetectionLocation highestLeftAnkle, highestRightAnkle, lowestLeftHeel, lowestRightHeel, lowestHighLeftHeel, lowestHighRightHeel;
    protected double MIN_JUMP = 0.04;
    int JUMP_LOOKBACK = 15;
    float scaler = 10;

    public PersonController(PersonState state) {
        super(state);
    }

    public int didJump(int lastJumpFrameID) {
        if (state.getInfoBlobArrayList().size() < JUMP_LOOKBACK) return 0;
        int lastFrameID = state.getInfoBlobArrayList().getLast().getFrameID();

        final int JUMP_FRAME_GAP = 2;

        UtilArrayList<DetectionLocation> leftHeels = state.leftLeg.heel.getNDetectedLocations(Math.min(JUMP_LOOKBACK, lastFrameID - lastJumpFrameID));
        UtilArrayList<DetectionLocation> rightHeels = state.rightLeg.heel.getNDetectedLocations(Math.min(JUMP_LOOKBACK, lastFrameID - lastJumpFrameID));
        lowestLeftHeel = leftHeels.getMin(DetectionLocation::compareByY);
        lowestRightHeel = rightHeels.getMin(DetectionLocation::compareByY);
        int leftHighIndex = leftHeels.indexOf(lowestLeftHeel);
        int rightHighIndex = rightHeels.indexOf(lowestRightHeel);

        // Lowest Ankle (by Position)
        UtilArrayList<DetectionLocation> leftAnkles = new UtilArrayList<>(leftHeels.size());
        leftAnkles.addAll(leftHeels.subList(0, leftHighIndex));
        UtilArrayList<DetectionLocation> rightAnkles = new UtilArrayList<>(rightHeels.size());
        rightAnkles.addAll(rightHeels.subList(0, rightHighIndex));
        highestLeftAnkle = leftAnkles.getMax(DetectionLocation::compareByY);
        highestRightAnkle = rightAnkles.getMax(DetectionLocation::compareByY);


        // Lowest Heel after Highest Heel (by Position)       => Downward Motion
        UtilArrayList<DetectionLocation> leftHighHeels = new UtilArrayList<>(leftHeels.size());
        leftHighHeels.addAll(leftHeels.subList(leftHighIndex + 1, leftHeels.size()));
        UtilArrayList<DetectionLocation> rightHighHeels = new UtilArrayList<>(rightHeels.size());
        rightHighHeels.addAll(rightHeels.subList(rightHighIndex + 1, rightHeels.size()));
        lowestHighLeftHeel = leftHighHeels.getMax(DetectionLocation::compareByY);
        lowestHighRightHeel = rightHighHeels.getMax(DetectionLocation::compareByY);
        //if the heel is lower(so actually higher...) than the ankle, we can consider this a jump


        if (
                (lowestLeftHeel != null && lowestRightHeel != null && lowestHighLeftHeel != null
                        && lowestHighRightHeel != null && highestLeftAnkle != null && highestRightAnkle != null)
                        && (
                        ((lowestLeftHeel.getY() + MIN_JUMP) < highestLeftAnkle.getY() && (lowestRightHeel.getY() + MIN_JUMP) < highestRightAnkle.getY()) &&
                                ((lowestHighLeftHeel.getY() - MIN_JUMP) > lowestLeftHeel.getY() && (lowestHighRightHeel.getY() - MIN_JUMP) > lowestRightHeel.getY())
                                && ((Math.max(lowestHighLeftHeel.getFrameID(), lowestHighRightHeel.getFrameID())) - lastJumpFrameID <= JUMP_FRAME_GAP)
                )
        ) {
            SLOG.d("DIDJUMP: " + lowestLeftHeel.getY() + " < " + highestLeftAnkle.getY() + " && " + lowestRightHeel.getY() + " < " + highestRightAnkle.getY() + " && " + lowestHighLeftHeel.getY() + " > " + lowestLeftHeel.getY() + " && " + lowestHighRightHeel.getY() + " > " + lowestRightHeel.getY());
            SLOG.d("DIDJUMP LastFrame: " + lastJumpFrameID + "  FrameID: " + Math.max(lowestHighLeftHeel.getFrameID(), lowestHighRightHeel.getFrameID()));

            return lastJumpFrameID;
        } else {
            return Math.max(lowestLeftHeel.getFrameID(), lowestRightHeel.getFrameID());
        }
    }

    public double getGroundLine() {

        UtilArrayList<DetectionLocation> leftAnkleLocations = state.leftLeg.ankle.getNDetectedLocations(GROUND_LOOKBACK);
        UtilArrayList<DetectionLocation> rightAnkleLocations = state.rightLeg.ankle.getNDetectedLocations(GROUND_LOOKBACK);

        return Math.max(leftAnkleLocations.getMax(DetectionLocation::compareByY).getY(), rightAnkleLocations.getMax(DetectionLocation::compareByY).getY());
    }

    public double getHeadLine() {

        UtilArrayList<DetectionLocation> leftEyeLocations = state.face.leftEye.getNDetectedLocations(HEAD_LOOKBACK);
        UtilArrayList<DetectionLocation> rightEyeLocations = state.face.rightEye.getNDetectedLocations(HEAD_LOOKBACK);

        return Math.min(leftEyeLocations.getMin(DetectionLocation::compareByY).getY(), rightEyeLocations.getMin(DetectionLocation::compareByY).getY());

    }

    public boolean inFrame() {
        return state.getDetectionSubClasses().stream().allMatch(ObjectDetectionState::inFrame);
    }

    public double getElbowFlexion(ArmState arm) {
        DetectionLocation shoulderLocation = arm.shoulder.getDetectedLocation();
        DetectionLocation elbowLocation = arm.elbow.getDetectedLocation();
        DetectionLocation wristLocation = arm.wrist.getLocation();
        return SMath.calculateAngle3Points(shoulderLocation, elbowLocation, wristLocation, false);
    }

    public double getLeftElbowFlexion() {
        return getElbowFlexion(state.leftArm);
    }

    public double getRighElbowFlexion() {
        return getElbowFlexion(state.rightArm);
    }

    public double getSideWaysExtendedCosine(ArmState arm) {
        //if the arm is extended, the cosine similarity between the shoulder-elbow and shoulder-wrist should be high
        DetectionLocation elbowD = arm.elbow.getDetectedLocation();
        DetectionLocation wristD = arm.wrist.getDetectedLocation();
        DetectionLocation shoulderD = arm.shoulder.getDetectedLocation();
        if (elbowD == null || wristD == null || shoulderD == null) return 0;
        //calculate the cosine similarity
        double shoulderX = shoulderD.getX();
        double shoulderY = shoulderD.getY();

        double scaledElbowX = (elbowD.getX() - shoulderX) * scaler;
        double scaledElbowY = (elbowD.getY() - shoulderY) * scaler;
        double scaledWristX = (wristD.getX() - shoulderX) * scaler;
        double scaledWristY = (wristD.getY() - shoulderY) * scaler;


        double xsim = (scaledElbowX) * (scaledWristX);
        double ysim = (scaledElbowY) * (scaledWristY);
        double enorm = Math.sqrt(Math.pow(scaledElbowX, 2) + Math.pow(scaledElbowY, 2));
        double wnorm = Math.sqrt(Math.pow(scaledWristX, 2) + Math.pow(scaledWristY, 2));

        double cosineSim = ((xsim + ysim) / (enorm * wnorm));
        SLOG.d("xsim: " + xsim + "ysim: " + ysim + "wnorm: " + wnorm + "enorm: " + enorm + "cosine: " + cosineSim);
        return cosineSim;

    }

    public boolean isleftArmSidewaysExtended() {
        return getSideWaysExtendedCosine(state.leftArm) > EXTENSION_THRESHOLD;
    }

    public boolean isrightArmSidewaysExtended() {
        return getSideWaysExtendedCosine(state.rightArm) > EXTENSION_THRESHOLD;
    }


    @Override
    public String getDebugText() {
        return "PersonController";
    }
}
